crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,130 @@
+//! Module to monitor changes in network.
+//! Get notified about that network path has changed
+//! and update local IP address cache
+use once_cell::sync::Lazy;
+use parking_lot::Mutex;
+use std::io;
+use std::sync::Arc;
+use telio_utils::{local_interfaces, telio_log_debug, telio_log_warn, GetIfAddrs};
+use tokio::{sync::broadcast::Sender, task::JoinHandle};
+/// Sender to notify if there is a change in OS interface order
+pub static PATH_CHANGE_BROADCAST: Lazy<Sender<()>> = Lazy::new(|| Sender::new(2));
+/// Vector containing all local interfaces
+pub static LOCAL_ADDRS_CACHE: Lazy<Arc<Mutex<Vec<if_addrs::Interface>>>> =
+    Lazy::new(|| Arc::new(Mutex::new(Vec::new())));
+#[cfg(all(
+    not(test),
+    any(target_os = "macos", target_os = "ios", target_os = "tvos")
+))]
+static NW_PATH_MONITOR_START: std::sync::Once = std::sync::Once::new();
+
+#[derive(Debug)]
+/// Struct to monitor network
+pub struct NetworkMonitor {
+    if_cache_updater_handle: Option<JoinHandle<io::Result<()>>>,
+    // Mac and Windows use callbacks, whereas in Linux a socket is registered
+    // with netlink multicast group messages
+    #[cfg(all(not(test), target_os = "linux"))]
+    monitor_handle: Option<JoinHandle<io::Result<()>>>,
+    #[cfg(all(not(test), target_os = "windows"))]
+    monitor_handle: crate::windows::SafeHandle,
+}
+
+impl NetworkMonitor {
+    /// Sets up and spawns network monitor
+    pub async fn new(if_addr: GetIfAddrs) -> std::io::Result<NetworkMonitor> {
+        *(LOCAL_ADDRS_CACHE.lock()) = local_interfaces::gather_local_interfaces(if_addr)?;
+
+        let if_cache_updater_handle = Some(tokio::spawn({
+            let mut notify = PATH_CHANGE_BROADCAST.subscribe();
+            async move {
+                loop {
+                    match notify.recv().await {
+                        Ok(()) => match local_interfaces::gather_local_interfaces(if_addr) {
+                            Ok(v) => {
+                                telio_log_debug!("Updating local addr cache");
+                                *(LOCAL_ADDRS_CACHE.lock()) = v;
+                            }
+                            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
+                        },
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        // Only add in production, otherwise network triggers from OS can occur
+        // which can make the test (test_path_change_notification) flaky sometimes
+        #[cfg(all(
+            not(test),
+            any(target_os = "macos", target_os = "ios", target_os = "tvos")
+        ))]
+        NW_PATH_MONITOR_START.call_once(crate::apple::setup_network_monitor);
```

Comment:
Out of curiosity, why not use `Once` also for Linux/Windows?

Reply:
It is not needed. I am not sure about Apple as well. I had just moved the Apple monitor. Earlier it was initiated from  `start_device`. Maybe it is just safer to add for Linux/Windows as well.

Reply:
Problem with adding `Once` for Linux/Windows is that it requires a return parameter. It is possible to return the monitor handle through some magic, but nothing straight forward so letting this be as is.